numba/core/typing/templates.py

```diff
@@ -720,7 +724,56 @@ def _build_impl(self, cache_key, args, kws):
             On failure, returns `(None, None)`.
 
         """
+        from numba.core import decorators
         from numba import jit
+        jitter_str = self.metadata.get('hardware', None)
+        if jitter_str is None:
+            # There is no hardware requested, use default, this preserves
+            # original behaviour
+            jitter = lambda *args, **kwargs: jit(*args, nopython=True, **kwargs)
+        else:
+            # Hardware has been requested, see what it is...
+            jitter = decorators.jit_registry.get(jitter_str, None)
+            from numba.core.extending_hardware import (hardware_registry,
+                                                       current_target)
+
+            def report_unknown_hardware(msg):
+                raise ValueError(msg.format(jitter_str))
+
+            if jitter is None:
+                # No JIT known for hardware string
+                hardware_class = hardware_registry.get(jitter_str, None)
+                if hardware_class is None:
+                    msg = ("Unknown hardware target '{}', has it been ",
+                           "registered?")
+                    report_unknown_hardware(msg)
+
+                # scan for the the current hardware target it's stuffed into the
+                # call stack at present, get that.
+                if len(self.context.callstack._stack) > 0:
+                    tos_target = self.context.callstack[0].target
+                else:
+                    tos_target = hardware_registry.get(current_target(), None)
```

Comment:
@sklam this is where the stack issues we discussed OOB stem from. The issue is essentially that if you enter the compilation chain from somewhere less standard then there could well be no frames and therefore no stack. This sort of thing being an example of said problem: https://github.com/numba/numba/pull/6898/files#diff-0e3c8738578e42578741958d965937324c13aaf8e6fc1523ce6affd318013dc3R592-R603